src/reflect/scala/reflect/internal/tpe/FindMembers.scala

```diff
@@ -13,8 +13,18 @@ trait FindMembers {
   this: SymbolTable =>
 
   /** Implementation of `Type#{findMember, findMembers}` */
-  private[internal] abstract class FindMemberBase[T](tpe: Type, name: Name, excludedFlags: Long, requiredFlags: Long) {
-    protected val initBaseClasses: List[Symbol] = tpe.baseClasses
+  private[internal] abstract class FindMemberBase[T](tpe: Type, name: Name, excludedFlags: Long, requiredFlags: Long, searchCompanions: Boolean) {
+    protected val initBaseClasses: List[Symbol] = {
+      val bcs = tpe.baseClasses
+      // For Java compilation units, a `T` or `A.T` can refer to a static inner class `T`. In the Scala compiler,
+      // we model static members are members of the companion object. Under `searchCompanions`, we always check the
+      // companion before proceeding to the next base class. This follows Java's semantics of "inheriting" static
+      // members: `A.T` can resolve to a static member defined in a parent of `A`.
+      //   class C { class T }; class B extends C { static class T }; class A extends B
+      // Here, `A.T` resolves to the static class `B.T`, so we need to check B's companion before proceeding to C.
+      if (searchCompanions) bcs.flatMap(bc => List(bc, bc.companionModule.moduleClass))
```

Comment:
Based on my example above, I think we also need to support the case where we're looking for the base classes of a module class. The companion modules of its companion class's base classes should also be included.